cf-agent/package_module.c

```diff
@@ -0,0 +1,1611 @@
+    for (Rlist *rp = options; rp != NULL; rp = rp->next)
+    {
+        char *value = RlistScalarValue(rp);
+        BufferAppendString(data, "options=");
+        BufferAppendString(data, value);
+        BufferAppendString(data, "\n");
+    }
+    return BufferClose(data);
+}
+
+static PackageInfo *ParseAndCheckPackageDataReply(const Rlist *data)
+{
+    PackageInfo * package_data = xcalloc(1, sizeof(PackageInfo));
+    
+    for (const Rlist *rp = data; rp != NULL; rp = rp->next)
+    {
+        char *line = RlistScalarValue(rp);
+                   
+        if (StringStartsWith(line, "PackageType="))
+        {
+            char *type = line + strlen("PackageType=");
+            if (StringSafeEqual(type, "file"))
+            {
+                package_data->type = PACKAGE_TYPE_FILE;
+            }
+            else if (StringSafeEqual(type, "repo"))
+            {
+                package_data->type = PACKAGE_TYPE_REPO;
+            }
+            else
+            {
+                Log(LOG_LEVEL_VERBOSE, "unsupported package type: %s", type);
+                free(package_data);
+                return NULL;
+            }
+        }
+        else if (StringStartsWith(line, "Name="))
+        {
+            if (package_data->name)
+            {
+                /* Some error occurred as we already have name for 
+                 * given package. */
+                Log(LOG_LEVEL_ERR, "Duplicated package name "
+                    "received: [%s] %s", line, package_data->name);
+                free(package_data);
+                return NULL;
+            }
+            package_data->name = 
+                SafeStringDuplicate(line + strlen("Name="));
+        }
+        else if (StringStartsWith(line, "Version="))
+        {
+            if (package_data->version)
+            {
+                /* Some error occurred as we already have version for 
+                 * given package. */
+                Log(LOG_LEVEL_ERR, "Duplicated package version "
+                    "received: [%s] %s", line, package_data->version);
+                free(package_data);
+                return NULL;
+            }
+            package_data->version = 
+                SafeStringDuplicate(line + strlen("Version="));
+        }
+        else if (StringStartsWith(line, "Architecture="))
+        {
+            if (package_data->arch)
+            {
+                /* Some error occurred as we already have arch for 
+                 * given package. */
+               Log(LOG_LEVEL_ERR, "Duplicated package architecture "
+                    "received: [%s] %s", line, package_data->arch);
+                free(package_data);
+                return NULL;
+            }
+            package_data->arch = 
+                SafeStringDuplicate(line + strlen("Architecture="));
+        }
+        /* For handling errors */
+        else if (StringStartsWith(line, "Error="))
+        {
+            Log(LOG_LEVEL_VERBOSE, "have error: %s",
```

Comment:
"have error" is not very helpful if you're reading the logs.  Maybe you can indicate the component name?